### DISCUSSION TITLE
Integrate Vue Flow into laravel nova

### DISCUSSION THREAD

**Discussion by scramatte** [bcakmakoglu/vue-flow#489] [Q&A] [answered]
Hi,

I'm trying to create a component for Laravel Nova that will use Vue Flow.
Nova use internally Vue3.  I've created a "Nova Tool" and I cut and paste your basic example.  I don't have any error on console .
What I've missed ? 

```
<template>
  <div>
    <Head title="Nova Workflow Designer" />

    <Heading class="mb-6">Nova Workflow Designer</Heading>

    <Card
      class="flex flex-col items-center justify-center"
      style="min-height: 600px"
    >
      <VueFlow v-model="elements" />
    </Card>
  </div>
</template>

<script>
import { ref } from "vue"
import { VueFlow  } from '@vue-flow/core'

export default {
  setup () {

    const elements = ref([
      { id: '1', type: 'input', label: 'Node 1', position: { x: 250, y: 5 } },
      { id: '2', label: 'Node 2', position: { x: 100, y: 100 }, },
      { id: '3', label: 'Node 3', position: { x: 400, y: 100 } },
      { id: '4', type: 'output', label: 'Node 4', position: { x: 400, y: 200 } },
      { id: 'e1-3', source: '1', target: '3' },
      { id: 'e1-2', source: '1', target: '2', animated: true },
    ]) 

    return {
      elements
    }
  }
}


</script>

<style>
@import '@vue-flow/core/dist/style.css'
</style>
```

**Comment by bcakmakoglu** [comment]
Can you try importing the default theme as well? `@import '@vue-flow/core/dist/theme-default.css';`

Also could you check if the VueFlow element is rendered into your DOM? (check with the browser devtools if the `div` with the class `vue-flow` exists.

**Comment by scramatte** [comment]
Still empty.
I attach a screenshot of the markup. Note that flow is embedded into a "Card"  has a flex design.  Not sure but it looks that size is 0x0 ..
![image](https://user-images.githubusercontent.com/188766/206174106-4c09d5fa-a0ee-4476-8c51-9e093b4a8b4a.png)

**Reply by bcakmakoglu** [reply]
Ah, you didn't add `VueFlow` to the component definitions, hence the tag isn't rendered properly :)
Change your component to this

```js
export default {
  components: { VueFlow },
  setup () {
    const elements = ref([
      { id: '1', type: 'input', label: 'Node 1', position: { x: 250, y: 5 } },
      { id: '2', label: 'Node 2', position: { x: 100, y: 100 }, },
      { id: '3', label: 'Node 3', position: { x: 400, y: 100 } },
      { id: '4', type: 'output', label: 'Node 4', position: { x: 400, y: 200 } },
      { id: 'e1-3', source: '1', target: '3' },
      { id: 'e1-2', source: '1', target: '2', animated: true },
    ]) 

    return {
      elements
    }
  }
}

**Reply by bcakmakoglu** [reply]
Using components without adding them to the `components: {}` definition only works if you're using the `<script setup>` syntax

**Comment by scramatte** [comment]
Now it looks that components is loaded but styles are not renderd

![image](https://user-images.githubusercontent.com/188766/206176112-bc1ed4a1-c8c0-41b4-8210-8ec24168b176.png)

I've make some change to the markup to have a container for Vue Flow. 

```
    <Card
      class="flex flex-col items-center justify-center"
      style="min-height: 600px;"
    > <div style="min-height: 600px; width: 100%; border: 1px solid red; margin: 10px;">
      <VueFlow v-model="elements" />
      </div>
    </Card>

```

**Reply by bcakmakoglu** [reply]
Well if you don't import the default theme, you'll have an unstyled Flow :)

**Reply by scramatte** [reply]
I believe that I've done it  

```
<style>
@import '@vue-flow/core/dist/style.css' 
@import '@vue-flow/core/dist/theme-default.css';
</style>
```

**Reply by bcakmakoglu** [reply]
Can you check the nodes in the dev-tools and see if the styles are present?
It looks like your styles simply aren't applied, maybe modify your import path to `node_modules/@vue-flow/core/dist/theme-default.css`

**Comment by scramatte** [comment]
Path is fine as  If I set "node_modules"  before I got webpack compilation error and I can see styles into dev-tools. 
If I remove **theme-default.css**  container is blank , I don't see the title of block

**Reply by bcakmakoglu** [reply]
Meaning? It works?
If not, can you show me the styles you see in the devtools? Like a screenshot of the devtools style tab?

**Reply by bcakmakoglu** [reply]
Also you can just use the respond function instead of opening a new Answer everytime you respond 😉 
Keeps the conversation cleaner in the threads

**Reply by scramatte** [reply]
Sorry, mistake on my side: Styles are not loaded. It looks that nova eat them. I've tried to inject them in various way and nothing.
Strangely, I've tried with another Vue3 component (https://n00ts.github.io/vue3-treeview/) and it works seamlessly.  

In script  or syle  block, compilation is fine ... but styles not available. 
I'm going to try to cut and past styles inside the component just to check.

```
<style scoped lang="css">
@import "@vue-flow/core/dist/style.css"
@import "@vue-flow/core/dist/theme-default.css"

</style>
```

**Reply by bcakmakoglu** [reply]
Don't use scoped for this purpose.
Scoped will only apply to elements that are statically analyzed inside your component, not for children etc.

Try adding the styles to your main style.css file or your app entry-point (like a main.ts file or whatever entry point you use with nova) instead.

**Reply by bcakmakoglu** [reply]
Are you possibly using something to purge css?
If so, you might want to add the vue flow css to it's ignore option, so that the styles are not purged from your output :)
As elements aren't statically available in your app and are created afterwards, the purge process might assume the classes are unused and thus just removes them from the css completely

**Reply by scramatte** [reply]
![image](https://user-images.githubusercontent.com/188766/206190184-e376e60b-e57a-416b-bea2-7157aee4c1e8.png)

Now I've got the styles but nothing change.

**Reply by bcakmakoglu** [reply]
Those are just the base styles, not the default theme.
Check the nodes for the styles, they probably have nothing applied to them.

Are you purging CSS?

**Reply by scramatte** [reply]
Might be a problem with z-index?

**Reply by bcakmakoglu** [reply]
I don't think so, no ^^
You see a canvas, and you seem to see your node labels. So no z-index issue. The main wrapper is supposed to be at 0 and it works with any other project I have used it in ;)

**Reply by bcakmakoglu** [reply]
If the Canvas was below your card, you wouldn't see any of your node labels, they'd also be stuck below your card, so I still have to assume it's simply the default theme that's not being applied.
Can you please check in the devtools for `vue-flow__node` elements and screenshot the styles for them? :)

**Reply by scramatte** [reply]
![image](https://user-images.githubusercontent.com/188766/206193948-8ff54efb-bc73-4568-a7b7-2ec726ae8f01.png)

Both stylesheet are loaded but no improvement ...

**Reply by scramatte** [reply]
purgeCss is not used as fare as I know

**Reply by bcakmakoglu** [reply]
Pretty sure they're not properly loaded still because here's what an example looks like if you simply remove the styles:
<img width="381" alt="image" src="https://user-images.githubusercontent.com/78412429/206195330-d876b824-1008-401c-8698-2e9ae04d89b0.png">

That's basically the same output that you have, so the issue is most certainly the styles not being applied 😢 

I asked before but can you show me the *dev-tools* tab for the styles of an element with the `vue-flow__node` class?

**Reply by bcakmakoglu** [reply]
You're using the latest TailwindCSS I assume, right?
I believe in versions below 3 they were using PurgeCSS but if you're at the latest version (i.e. TailwindCSS 3+) there's no purgeCSS anymore.

Something still does seem to interfere with the styles though

**Reply by scramatte** [reply]
It looks that tailwindcss 3.2.0  is used

**Reply by bcakmakoglu** [reply]
Can you try copy + pasting the styles from the vue flow css files and just putting them into your style tag and see if that makes any difference?
Obviously that's not the way to do it but just for testing purposes I'd like to know if that works ^^

**Reply by scramatte** [reply]
It doesn't works. I've just send an issue to the nova-issues GitHub, I hope that they light us ...
https://github.com/laravel/nova-issues/issues/5141

**Reply by bcakmakoglu** [reply]
mh... copy pasting the styles didn't work either?
That's very strange :/

I've asked before but can you give me a screenshot of the dev-tools style tab when checking a `vue-flow__node` element?
That would make it clear if the styles are really not there :)

Just open the devtools, search for `vue-flow__node` and show me the devtools right panel where the element styles are listed so I can see if they exist.
If they do, there must be something else going wrong but I want to make sure it *is 100%* that issue and nothing else.

**Reply by scramatte** [reply]
![image](https://user-images.githubusercontent.com/188766/206243959-0443a619-69b6-41a6-8837-2a29bff4c687.png)

It doesn't exists...

**Reply by bcakmakoglu** [reply]
You have to search for the element ^^
It looks like you're on the `vue-flow__viewport` element.

<img width="565" alt="image" src="https://user-images.githubusercontent.com/78412429/206246933-11be9d77-2040-40ac-bfac-f3d7e57e53b2.png">

It should look like this

**Reply by scramatte** [reply]
I've added CSS into PHP Laravel Provider as explained in Nova manual. And now I'got the styles but look that everything is hidden ...

![image](https://user-images.githubusercontent.com/188766/206252497-16d1ea72-2056-4e58-9665-5a94a4228b73.png)
![image](https://user-images.githubusercontent.com/188766/206253018-203d0dc1-1189-4010-8ed6-d06d748afe46.png)

**Reply by scramatte** [reply]
It works !
![image](https://user-images.githubusercontent.com/188766/206254712-606106ad-8b9d-48c6-8266-75da1caf0e66.png)

The problem was first the CSS and in second time you must specify   height of the container.
It looks that the height of  vue-flow__container is not calculated properly.   I understand that with min-height it should be enough but not ... maybe you can fix it?

```
    <Card
      class="flex flex-col items-center justify-center"
      style="min-height: 600px; height: 600px;"
    >
       <VueFlow v-model="elements" />
    </Card>
```

**Reply by bcakmakoglu** [reply]
Vue Flow doesn't calculate any height, it just takes the height from it's wrapping element, there's no fixing that ^^

**Comment by bcakmakoglu** [comment, marked as answer]
Just so other users find the answer to this question: You *need* to set proper dimensions for the wrapping element of VueFlow, meaning it needs to have a width and a height, otherwise VueFlow will not work properly.
